### ISSUE TITLE
Bundler 2.5.17 breaks bundling vendored gems from git, when git is not installed `during bundle install --local`

### ISSUE THREAD

**Issue by sethboyles** [rubygems/rubygems#8217]
### Describe the problem as clearly as you can

In previous versions of Bundler, if I vendor my gems like so:

```
bundle config set cache_all true
bundle __BUNDLER_VERSION_  package --all-platforms --no-install
```

then later, on a different machine without `git` available, when installing from a local cache (`bundle install --local`), this would work.  On 2.5.17 and above, I am met with this error:

```
Fetching https://github.com/sporkrb/spork
You need to install git to be able to use gems from git repositories. For help
installing git, please refer to GitHub's tutorial at
https://help.github.com/articles/set-up-git
```

I believe this is related to https://github.com/rubygems/rubygems/pull/4469 (cc @deivid-rodriguez), but I'm not sure how since that PR seems to test explicitly for this use-case.

### Did you try upgrading rubygems & bundler?

I tried with bundler 2.5.23 (released yesterday as of this posting), no luck

### Post steps to reproduce the problem

Here is a repo containing a Dockerfile that will reproduce the problem: https://github.com/sethboyles/bundle-no-git

Clone the repo and run `./rebuild-and-run.bash` to observe how `bundle _2.5.17_ install --local` without git available fails and `bundle _2.5.16_ install --local` without git succeeds.

### Which command did you run?

`bundle _2.5.17_ install --local`

### What were you expecting to happen?

Installation succeeds when git is not available.

### What actually happened?

I get the `You need to install git ` error

On 2.5.17:

```
Bundling with 2.5.17
Fetching https://github.com/sporkrb/spork
You need to install git to be able to use gems from git repositories. For help
installing git, please refer to GitHub's tutorial at
https://help.github.com/articles/set-up-git
```

On 2.5.16:

```
Bundling with 2.5.16
Updating files in vendor/cache
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
```

### If not included with the output of your command, run `bundle env` and paste the output below

https://gist.github.com/sethboyles/2019320da3c01cae1773f71a5844f4e4

**Comment by deivid-rodriguez** [comment]
Hello, thanks for the fantastic report!

Unfortunately, this is an unintended side effect the PR you mentioned, which was meant to fix caching of git repositories. Before, the cache folder for git repositories would be just a checkout of the git repository, so "installing" git repositories in `--local` mode was just a matter of copying the cache folder to the installation folder (not git necessary). However, that didn't work for gems with extensions, since extensions would never get actually compiled.

Now, we keep a bare clone of the git repository in the cache, and check out the bare clone into the installation folder when installing. But that unintentionally reintroduce the need for `git` to install git gems, even when you have `vendor/cache`.

We indeed have a spec checking exactly for this case, that's why this went unnoticed:

https://github.com/rubygems/rubygems/blob/da0faf964ce4eebba34bb624d05176e440050886/bundler/spec/cache/git_spec.rb#L271-L288

The problem is that the `simulate_new_machine` helper does not do what I expected, since it only removes the `.bundle` folder, but not the installation folder (`vendor/bundle`). So of course git is not necessary in that situation because gems are already installed. In #4469 I actually changed that spec to configure the installation path to `vendor/bundle`, so maybe I did detect a failure due to this but I failed to recognize it as a breaking change. I honestly don't recall 😞.

A potential solution would be to go back to keeping a checkout of the repository. I'll think about how feasible that is. The new bare cache format has caused a few issues, so maybe that's a good thing, but it will be tricky to handle backwards compatibility for people who already migrated to the new format (and most likely annoying for them anyways).
